File: mobile/app/src/main/java/ua/nure/cleaningservice/ui/rva/ServicesRVA.java
package ua.nure.cleaningservice.ui.rva;

import android.content.Context;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.cardview.widget.CardView;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;
import java.util.Locale;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import ua.nure.cleaningservice.R;
import ua.nure.cleaningservice.data.Company;
import ua.nure.cleaningservice.data.Service;
import ua.nure.cleaningservice.network.JSONPlaceHolderApi;
import ua.nure.cleaningservice.network.NetworkService;
import ua.nure.cleaningservice.ui.edit.EditServicesActivity;

public class ServicesRVA extends RecyclerView.Adapter<ServicesRVA.ServicesViewHolder>{

    private Context mContext;
    private ServicesViewHolder mServicesViewHolder;
    private JSONPlaceHolderApi mApi;
    String token;


    public ServicesViewHolder onCreateViewHolder(ViewGroup viewGroup, int i) {
        View v = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.services_card, viewGroup, false);
        mApi = NetworkService.getInstance().getApiService();
        token = "Bearer " + Company.getInstance().getToken();
        ServicesViewHolder servicesViewHolder = new ServicesViewHolder(v);
        return servicesViewHolder;
    }

    private List<Service> mServices;

    public ServicesRVA(Context context, List<Service> servicesList){
        this.mContext = context;
        this.mServices = servicesList;
    }

    @Override
    public void onBindViewHolder(@NonNull ServicesViewHolder holder, int position) {
        holder.mCardView.setId(mServices.get(position).getId());
        holder.mNameTV.setText(mServices.get(position).getName());
        holder.mDescTV.setText(mServices.get(position).getDescription());
        holder.mMinAreaTV.setText(String.format(Locale.getDefault(), "%d",mServices.get(position).getMinArea()));
        holder.mMaxAreaTV.setText(String.format(Locale.getDefault(), "%d",mServices.get(position).getMaxArea()));
        holder.mRoomTypeTV.setText(mServices.get(position).getPlacementType());
        holder.mPriceTV.setText(String.format(Locale.getDefault(), "%f", mServices.get(position).getPricePerMeter()));
        holder.mEditButton.setOnClickListener(v -> {
            editServices(holder);
        });
        holder.mDeleteButton.setOnClickListener(v -> {
            mServicesViewHolder = holder;
            deleteService();
        });
    }
    @Override
    public void onAttachedToRecyclerView(RecyclerView recyclerView) {
        super.onAttachedToRecyclerView(recyclerView);
    }

    @Override
    public int getItemCount() {
        return mServices.size();
    }

    class ServicesViewHolder extends RecyclerView.ViewHolder {
        CardView mCardView;
        TextView mNameTV, mDescTV, mMinAreaTV, mMaxAreaTV, mRoomTypeTV, mPriceTV;
        Button mEditButton, mDeleteButton;
        ServicesViewHolder(View itemView) {
            super(itemView);
            mCardView = (CardView) itemView.findViewById(R.id.services_cv);
            mNameTV = (TextView) itemView.findViewById(R.id.name_service_label);
            mDescTV = (TextView) itemView.findViewById(R.id.desc_text);
            mMinAreaTV = (TextView) itemView.findViewById(R.id.min_area_text);
            mMaxAreaTV = (TextView) itemView.findViewById(R.id.max_area_text);
            mRoomTypeTV = (TextView) itemView.findViewById(R.id.room_type_text);
            mPriceTV = (TextView) itemView.findViewById(R.id.price_text);
            mEditButton = (Button) itemView.findViewById(R.id.edit_service_btn);
            mDeleteButton = (Button) itemView.findViewById(R.id.delete_service_btn);
        }

    }

    private void editServices(ServicesViewHolder holder) {
        Intent intent = new Intent(mContext, EditServicesActivity.class);
        intent.putExtra("sId", holder.mCardView.getId());
        mContext.startActivity(intent);
    }

    private void deleteService() {
        mApi.deleteService(token, mServicesViewHolder.mCardView.getId()).enqueue(deleteCallback);
    }
    Callback<Service> deleteCallback = new Callback<Service>() {
        @Override
        public void onResponse(Call<Service> call, Response<Service> response) {
            if (response.isSuccessful()) {
                System.out.println(response.body());
                mServices.remove(mServicesViewHolder.getAdapterPosition());
                notifyDataSetChanged();
            }
        }

        @Override
        public void onFailure(Call<Service> call, Throwable t) {
            System.out.println(t);
        }
    };

}

File: mobile/app/src/main/java/ua/nure/cleaningservice/ui/rva/ServicesRVA.kt
package ua.nure.cleaningservice.ui.rva

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.cardview.widget.CardView
import androidx.recyclerview.widget.RecyclerView
import ua.nure.cleaningservice.R
import ua.nure.cleaningservice.data.Service
import ua.nure.cleaningservice.data.User
import ua.nure.cleaningservice.network.JsonPlaceHolderApi
import ua.nure.cleaningservice.network.NetworkService
import ua.nure.cleaningservice.ui.rva.ServicesRVA.ServicesViewHolder


class ServicesRVA(val mServices: MutableList<Service>?) : RecyclerView.Adapter<ServicesViewHolder>() {

    private var mApi: JsonPlaceHolderApi? = null
    var token: String? = null

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): ServicesViewHolder {
        val v = LayoutInflater.from(viewGroup.context).inflate(R.layout.services_card, viewGroup, false)
        mApi = NetworkService.getInstance().apiService
        token = "Bearer " + User.getInstance().token
        return ServicesViewHolder(v)
    }

    override fun onBindViewHolder(holder: ServicesViewHolder, position: Int) {
        val name = mServices?.get(position)?.name
        val id = mServices!![position].id
        val nameAndId = "$id - $name"
        holder.mCardView.id = id
        holder.mNameTV.text = nameAndId
        holder.mDescTV.text = mServices[position].description
        holder.mMinAreaTV.text = mServices[position].minArea.toString()
        holder.mMaxAreaTV.text = mServices[position].maxArea.toString()
        holder.mRoomTypeTV.text = mServices[position].placementType
        holder.mPriceTV.text = mServices[position].pricePerMeter.toString()
    }

    override fun getItemCount(): Int {
        return mServices!!.size
    }

    inner class ServicesViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var mCardView: CardView = itemView.findViewById<View>(R.id.services_cv) as CardView
        var mNameTV: TextView = itemView.findViewById<View>(R.id.name_service_label) as TextView
        var mDescTV: TextView = itemView.findViewById<View>(R.id.desc_text) as TextView
        var mMinAreaTV: TextView = itemView.findViewById<View>(R.id.min_area_text) as TextView
        var mMaxAreaTV: TextView = itemView.findViewById<View>(R.id.max_area_text) as TextView
        var mRoomTypeTV: TextView = itemView.findViewById<View>(R.id.room_type_text) as TextView
        var mPriceTV: TextView = itemView.findViewById<View>(R.id.price_text) as TextView
    }
}
